torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional, Tuple
+
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+
+class PackedDataset(Dataset):
+    """
+    Performs greedy sample packing on a provided dataset. This is done as a single
+    preprocessing step before training begins. Shuffling is done outside of this
+    class on packed samples as part of the dataloader.
+
+    The class loads, tokenizes, and packs examples on initialization - no tokenization is done during training.
+
+    The general flow on initialization is: load tokenized sample -> add to buffer ->
+        when buffer is long enough, add to self.samples.
+
+    During training, returns self.samples[idx] as input and label.
+
+    Args:
+        ds (Dataset): dataset to sample pack. This should return a tuple of tokenized
+            inputs and labels.
+        max_seq_len (int): Maximum number of tokens to pack
+        max_rows (Optional[int]): maximum number of samples to pack. Default is None, which will pack as many samples as possible.
+    """
+
+    def __init__(
+        self,
+        ds: Dataset,
+        max_seq_len: int,
+        max_rows: Optional[int] = None,
+    ) -> None:
+        self.ds = ds
+        self.max_seq_len = max_seq_len
+        self.max_rows = max_rows
+        # where final samples will be held
+        self.samples: List[Dict[str, List[int]]] = []
+        self._pack()
+
+    def _pack(self) -> None:
+        """
+        Iterate through the dataset. Use a buffer to hold samples until max_seq_len,
+        then append the buffer to self.samples as a single "packed" sample. Continue
+        until max_rows or end of dataset.
+        """
+        # buffer to hold samples until they are long enough to be added to self.samples
+        buffer = {
+            "input_ids": [],
+            "labels": [],
```

Comment:
I am actually thinking about whether we should also build attention/sentence masks for packed sequence (something like flash-attn did: https://github.com/Dao-AILab/flash-attention/issues/432#issuecomment-1698610752). So that the model would be able to know which part belongs to sentence 1, which part belongs to sentence 2, etc, this can be used to adjust the self-attention so that different sequences cannot attend to each other.

Reply:
Yes, absolutely agree. Having an integer mask that assigns an integer for each sample in the pack to prevent cross-attending should be supported. Although, I need to think about how to propagate this mask properly to the model forward, right now our datasets have only returned the input tokens and labels. But this should probably change...

Reply:
This is indeed something we may want, but... the guys from trl have been doing packing for a while without this feature with really good fine-tunes.

Reply:
@tcapelle you mean without masking or something else? If it's without masking, that's just.....wrong? Even if it gives decent performance on benchmarks, I dont think we want to enable that?